### PR TITLE
ARROW-13542: [C++][Compute][Dataset] Add dataset::WriteNode for writing rows from an ExecPlan to disk

### PR DESCRIPTION
**Pull request by westonpace**
This PR adds a write node.  The write node takes in `FileSystemDatasetWriteOptions` and the projected schema and writes the incoming data to disk.  It is a sink node but it is a bit different than the existing sink node.  The existing sink node transferred batches via an AsyncGenerator but that puts ownership of the batches outside of the push-based flow of the exec plan.  I added a new ConsumingSinkNode which consumes the batches as part of the push-based flow.  This makes it possible to block the exec plan from finishing until all data has been written to disk.

In addition this PR refines the AsyncTaskGroup a little.  `WaitForTasksToFinish` was not a very clearly named method.  Once called it actually transitioned the task group from a "top level tasks can be added" state to a "no more top level tasks can be added" state and that was not clear in the name.  The new name (`End`) is hopefully more clear.

This PR does not solve the backpressure problem.  Instead a serial async task group was created.  This will run all tasks in order (the default task group allows them to run in parallel) but not necessarily on the calling thread (i.e. unlike SerialTaskGroup we do not block on the AddTask method).  This allows tasks to pile up in a queue and, if the write is slow, this will become a pile-up point which will eventually run out of memory (provided there is enough data being written).  That problem is solved in a follow-up #11286 

The `AsyncSerializedTaskGroup` and `AsyncTaskGroup` classes have very similar APIs but I did not create an interface / abstract base class because I don't yet envision any case where they would be interchangeable.  The distinction is a "can these tasks run in parallel or not" and is not a performance / resource question.

As a consequence of using the ExecPlan dataset writes no longer have reliable ordering.  If you pass in three batches and they are all destined for the same file then the batches may be written in any order in the destination file.  This is because the ExecPlan creates a thread task for each input batch and so they could arrive at the write node in any order.

### CODE REVIEW

File: cpp/src/arrow/compute/exec/sink_node.cc [hunk 1]
```diff
@@ -132,6 +133,104 @@ class SinkNode : public ExecNode {
   PushGenerator<util::optional<ExecBatch>>::Producer producer_;
 };
 
+// A sink node that owns consuming the data and will not finish until the consumption
+// is finished.  Use SinkNode if you are transferring the ownership of the data to another
+// system.  Use ConsumingSinkNode if the data is being consumed within the exec plan (i.e.
+// the exec plan should not complete until the consumption has completed).
+class ConsumingSinkNode : public ExecNode {
```
Comment (lidavidm) [inline cpp/src/arrow/compute/exec/sink_node.cc:140]: Should we add an explicit test of this node in plan_test.cc, as with the other nodes?

**Review by lidavidm — APPROVED** [review]
LGTM. There are a lot of files touched, but things look straightforward.

File: cpp/src/arrow/util/future.h [hunk 2]
```diff
@@ -840,6 +840,17 @@ inline Future<>::Future(Status s) : Future(internal::Empty::ToResult(std::move(s
 ARROW_EXPORT
 Future<> AllComplete(const std::vector<Future<>>& futures);
 
+/// \brief Create a Future which completes when all of `futures` complete.
+///
+/// The future will be marked complete if all `futures` complete
+/// successfully. Otherwise, it will be marked failed with the status of
+/// the first failing future.
```
[the remark this answers is not in the document]
Comment (westonpace) [reply cpp/src/arrow/util/future.h]: I clarified the comment according to that review and also changed the name to AllFinished.

File: cpp/src/arrow/compute/exec/sink_node.cc (hunk 1 above)
Comment (westonpace) [reply cpp/src/arrow/compute/exec/sink_node.cc:140]: I added a few unit tests to plan_test.cc

File: cpp/src/arrow/compute/exec/plan_test.cc [hunk 3]
```diff
@@ -374,6 +374,108 @@ TEST(ExecPlanExecution, SourceSinkError) {
...
+  for (bool slow : {false, true}) {
+    SCOPED_TRACE(slow ? "slowed" : "unslowed");
+
+    for (bool parallel : {false, true}) {
+      SCOPED_TRACE(parallel ? "parallel" : "single threaded");
+      ASSERT_OK_AND_ASSIGN(auto plan, ExecPlan::Make());
+      uint32_t batches_seen = 0;
+      Future<> finish = Future<>::Make();
+      struct TestConsumer : public SinkNodeConsumer {
+        TestConsumer(uint32_t* batches_seen, Future<> finish)
```
Comment (lidavidm) [inline cpp/src/arrow/compute/exec/plan_test.cc:387]: In the parallel case, shouldn't this be an atomic?

Comment (lidavidm) [reply cpp/src/arrow/compute/exec/plan_test.cc:387]: (I suppose with only 2 batches, it's unlikely to run into this, but…)

**Review by lidavidm — APPROVED** [review]
Thanks for the updates. LGTM

File: cpp/src/arrow/compute/exec/plan_test.cc (hunk 3 above)
Comment (westonpace) [reply cpp/src/arrow/compute/exec/plan_test.cc]: Done.

**Pull request closed without merging**